src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs

```diff
@@ -2,161 +2,442 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using System.Diagnostics;
+using System.Linq;
 using System.IO;
 using System.Net.Test.Common;
 using System.Threading;
 using System.Threading.Tasks;
-
 using Xunit;
-using Xunit.Abstractions;
 
 namespace System.Net.Http.Functional.Tests
 {
-    public class CancellationTest : HttpClientTestBase
+    public class HttpClientHandler_CancellationTest : HttpClientTestBase
     {
-        private readonly ITestOutputHelper _output;
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(TwoBoolsAndCancellationMode))]
+        public async Task PostAsync_CancelDuringRequestContentSend_TaskCanceledQuickly(bool chunkedTransfer, bool connectionClose, CancellationMode mode)
+        {
+            if (IsWinHttpHandler || IsNetfxHandler)
+            {
+                // hangs / doesn't cancel
+                return;
+            }
+
+            using (HttpClient client = CreateHttpClient())
+            {
+                client.Timeout = Timeout.InfiniteTimeSpan;
+                var cts = new CancellationTokenSource();
+
+                await LoopbackServer.CreateServerAsync(async (server, url) =>
+                {
+                    Task serverTask = LoopbackServer.AcceptSocketAsync(server, async (socket, stream, reader, writer) =>
+                    {
+                        // Since we won't receive all of the request, just read everything we do get
+                        byte[] ignored = new byte[100];
+                        while (await stream.ReadAsync(ignored, 0, ignored.Length) > 0);
+                        return null;
+                    });
+
+                    var preContentSent = new TaskCompletionSource<bool>();
+                    var sendPostContent = new TaskCompletionSource<bool>();
+
+                    await ValidateClientCancellationAsync(async () =>
+                    {
+                        var req = new HttpRequestMessage(HttpMethod.Post, url);
+                        req.Content = new DelayedByteContent(2000, 3000, preContentSent, sendPostContent.Task);
+                        req.Headers.TransferEncodingChunked = chunkedTransfer;
+                        req.Headers.ConnectionClose = connectionClose;
+
+                        Task<HttpResponseMessage> postResponse = client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, cts.Token);
+                        await preContentSent.Task;
+                        if ((mode & CancellationMode.Token) != 0) cts.Cancel();
```

Comment:
Nit: since you do this CancellationMode "switch" several places, you could factor it into a helper function to ensure consistency and simplify the code a bit.

Reply:
Yup